File: tests/suites/createPair.ts
import { expect } from "chai";
import {
  setProvider,
  Provider,
  Program,
  workspace,
  BN,
} from "@project-serum/anchor";
import {
  PublicKey,
  Keypair,
  SystemProgram,
  SYSVAR_RENT_PUBKEY,
} from "@solana/web3.js";
import { Pools } from "../../target/types/pools";
import { airdropUsers } from "../helpers";
import {
  AccountInfo,
  ASSOCIATED_TOKEN_PROGRAM_ID,
  Token,
  TOKEN_PROGRAM_ID,
} from "@solana/spl-token";

export const testCreatePair = (provider: Provider) =>
  describe("Create a liquidity pair", () => {
    setProvider(provider);

    const program = workspace.Pools as Program<Pools>;

    let creator: Keypair;
    let tokenA: Token, tokenB: Token;
    let accountA: AccountInfo, accountB: AccountInfo;
    let id = new BN(5);
    const initialSupply = new BN(10 ** 10);
    const baseRewards = new BN(10 ** 10);
    const timeMultiplier = new BN(192);

    before(async () => {
      creator = Keypair.generate();
      await airdropUsers([creator], provider);

      tokenA = await Token.createMint(
        provider.connection,
        creator,
        creator.publicKey,
        null,
        9,
        TOKEN_PROGRAM_ID
      );
      accountA = await tokenA.getOrCreateAssociatedAccountInfo(
        creator.publicKey
      );
      await tokenA.mintTo(
        accountA.address,
        creator,
        [],
        initialSupply.toNumber()
      );
      tokenB = await Token.createMint(
        provider.connection,
        creator,
        creator.publicKey,
        null,
        9,
        TOKEN_PROGRAM_ID
      );
      accountB = await tokenB.getOrCreateAssociatedAccountInfo(
        creator.publicKey
      );
      await tokenB.mintTo(
        accountB.address,
        creator,
        [],
        initialSupply.toNumber()
      );
    });

    it("Create a liquidity pair", async () => {
      const [pairAddress, pairBump] = await PublicKey.findProgramAddress(
        [
          Buffer.from("pair"),
          tokenA.publicKey.toBuffer(),
          tokenB.publicKey.toBuffer(),
        ],
        program.programId
      );
      const [authorityAddress, authorityBump] =
        await PublicKey.findProgramAddress(
          [
            Buffer.from("authority"),
            tokenA.publicKey.toBuffer(),
            tokenB.publicKey.toBuffer(),
          ],
          program.programId
        );
      const [mintAddress, mintBump] = await PublicKey.findProgramAddress(
        [
          Buffer.from("mint"),
          tokenA.publicKey.toBuffer(),
          tokenB.publicKey.toBuffer(),
        ],
        program.programId
      );
      const [accountAAddress, accountABump] =
        await PublicKey.findProgramAddress(
          [
            Buffer.from("account_a"),
            tokenA.publicKey.toBuffer(),
            tokenB.publicKey.toBuffer(),
          ],
          program.programId
        );
      const [accountBAddress, accountBBump] =
        await PublicKey.findProgramAddress(
          [
            Buffer.from("account_b"),
            tokenA.publicKey.toBuffer(),
            tokenB.publicKey.toBuffer(),
          ],
          program.programId
        );
      const [feesAAddress, feesABump] = await PublicKey.findProgramAddress(
        [
          Buffer.from("fees_a"),
          tokenA.publicKey.toBuffer(),
          tokenB.publicKey.toBuffer(),
        ],
        program.programId
      );
      const [feesBAddress, feesBBump] = await PublicKey.findProgramAddress(
        [
          Buffer.from("fees_b"),
          tokenA.publicKey.toBuffer(),
          tokenB.publicKey.toBuffer(),
        ],
        program.programId
      );

      const bumps = {
        pair: pairBump,
        authority: authorityBump,
        mint: mintBump,
        accountA: accountABump,
        accountB: accountBBump,
        feesA: feesABump,
        feesB: feesBBump,
      };
      const stable = true;

      await program.rpc.createPair(bumps, stable, {
        accounts: {
          pair: pairAddress,
          authority: authorityAddress,
          pairMint: mintAddress,
          mintA: tokenA.publicKey,
          mintB: tokenB.publicKey,
          pairAccountA: accountAAddress,
          pairAccountB: accountBAddress,
          feesAccountA: feesAAddress,
          feesAccountB: feesBAddress,
          creator: creator.publicKey,
          tokenProgram: TOKEN_PROGRAM_ID,
          rent: SYSVAR_RENT_PUBKEY,
          systemProgram: SystemProgram.programId,
        },
        signers: [creator],
      });

      const p = await program.account.pair.fetch(pairAddress);

      expect(p.mintA.toString()).to.equal(tokenA.publicKey.toString());
      expect(p.mintB.toString()).to.equal(tokenB.publicKey.toString());
      expect(p.stable).to.equal(stable);
      expect(p.pairMint.toString()).to.equal(mintAddress.toString());
      expect(p.authority.toString()).to.equal(authorityAddress.toString());
    });
  });

File: programs/pools/src/instructions/create_pair.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{self, Mint, Token, TokenAccount};

use crate::state::{CreatePairBumps, Pair};

#[derive(Accounts)]
#[instruction(bumps: CreatePairBumps)]
pub struct CreatePair<'info> {
    /// The pair
    #[account(
        init,
        payer = creator,
        seeds = [
            b"pair",
            mint_a.key().as_ref(),
            mint_b.key().as_ref(),
        ],
        bump = bumps.pair
    )]
    pub pair: Box<Account<'info, Pair>>,

    /// The pair authority
    #[account(
        mut,
        seeds = [
            b"authority",
            mint_a.key().as_ref(),
            mint_b.key().as_ref()
        ],
        bump = bumps.authority
    )]
    pub authority: AccountInfo<'info>,

    /// The mint of the token representing liquidity
    #[account(
        init,
        payer = creator,
        seeds = [
            b"mint",
            mint_a.key().as_ref(),
            mint_b.key().as_ref(),
        ],
        bump = bumps.mint,
        mint::decimals = 0,
        mint::authority = authority
    )]
    pub pair_mint: Box<Account<'info, Mint>>,

    /// The mint of the token A
    pub mint_a: AccountInfo<'info>,

    /// The mint of the token B
    pub mint_b: AccountInfo<'info>,

    /// The pair account holding token A
    #[account(
        init,
        payer = creator,
        seeds = [
            b"account_a",
            mint_a.key().as_ref(),
            mint_b.key().as_ref()
        ],
        bump = bumps.account_a,
        token::mint = mint_a,
        token::authority = authority
    )]
    pub pair_account_a: Box<Account<'info, TokenAccount>>,

    /// The pair account holding token B
    #[account(
        init,
        payer = creator,
        seeds = [
            b"account_b",
            mint_a.key().as_ref(),
            mint_b.key().as_ref()
        ],
        bump = bumps.account_b,
        token::mint = mint_b,
        token::authority = authority
    )]
    pub pair_account_b: Box<Account<'info, TokenAccount>>,

    /// The account holding fees on token A
    #[account(
        init,
        payer = creator,
        seeds = [
            b"fees_a",
            mint_a.key().as_ref(),
            mint_b.key().as_ref()
        ],
        bump = bumps.fees_a,
        token::mint = mint_a,
        token::authority = authority
    )]
    pub fees_account_a: Account<'info, TokenAccount>,

    /// The account holding fees on token B
    #[account(
        init,
        payer = creator,
        seeds = [
            b"fees_b",
            mint_a.key().as_ref(),
            mint_b.key().as_ref()
        ],
        bump = bumps.fees_b,
        token::mint = mint_b,
        token::authority = authority
    )]
    pub fees_account_b: Box<Account<'info, TokenAccount>>,

    /// The creator of the pool
    #[account(mut)]
    pub creator: Signer<'info>,

    /// The program for interacting with the token.
    #[account(address = token::ID)]
    pub token_program: Program<'info, Token>,

    pub rent: Sysvar<'info, Rent>,
    pub system_program: Program<'info, System>,
}

pub fn handler(ctx: Context<CreatePair>, bumps: CreatePairBumps, stable: bool) -> ProgramResult {
    let pair = &mut ctx.accounts.pair;
    pair.mint_a = ctx.accounts.mint_a.key();
    pair.mint_b = ctx.accounts.mint_b.key();
    pair.stable = stable;
    pair.pair_mint = ctx.accounts.pair_mint.key();
    pair.authority = ctx.accounts.authority.key();
    pair.bumps = bumps;

    msg!("Created pair for token A={} and B={}", pair.mint_a, pair.mint_b);

    Ok(())
}
